share/lib/python/neuron/hclass35.py

```diff
@@ -0,0 +1,40 @@
+#Python 3.0-3.5 only
+# ------------------------------------------------------------------------------
+# class factory for subclassing h.anyclass
+# h.anyclass methods may be overridden. If so the base method can be called
+# using the idiom self.basemethod = self.baseattr('methodname')
+# ------------------------------------------------------------------------------
+
+from neuron import h, hoc
+import nrn
+
+#avoid syntax error if compiled by python 2
+
+class MetaHocObject(type):
+  """Provides Exception for Inheritance of multiple HocObject"""
+  def __new__(cls, name, bases, attrs):
+    #print cls, name, bases
+    m = []
+    for b in bases:
+      if issubclass(b, hoc.HocObject):
+        m.append(b.__name__)
+    if (len(m) > 1):
+      raise TypeError('Multiple Inheritance of HocObject in %s' % name
+        + ' through %s not allowed' % ','.join(m))
+      #note that join(b.__name__ for b in m) is not valid for Python 2.3
+
+    return type.__new__(cls, name, bases, attrs)
+
+def hclass(c):
+    """Class factory for subclassing h.anyclass. E.g. class MyList(hclass(h.List)):..."""
+    if c == h.Section :
+        return nrn.Section
+    class hc(hoc.HocObject, metaclass=MetaHocObject):
+    #class hc(hoc.HocObject):
+        def __new__(cls, *args, **kwds):
+            kwds2 = {'hocbase': cls.htype}
+            if 'sec' in kwds:
+                kwds2['sec'] = kwds['sec']
+            return hoc.HocObject.__new__(cls, *args, **kwds2)
+    setattr(hc, 'htype', c)
+    return hc
```

Comment:
This is all outside my python competence but that was pretty much the case also when this was originally implemented.
So I accept the improvement. And Thanks!

Originally, on first implementation, there was no neuron module, just hoc and nrn. After some use it was desirable that
hoc be the goto place for everything and the nrn module is now only a place for the few other objects and modules it manages such as nrn.Section. It may help to realize that all the methods that seem like constructors of hoc class instances```h.Section()```, ```h.Vector()```, ```h.IClamp(nrn.segment)```, et many cetera, are actually method factories to construct the objects. Section has always been a vexing case as in the hoc world it is not already an object  and a lot of shenanigans are going on to make it look like an object in python. Probably the most difficult to
deal with issues have to do with the Section name and keeping the hoc and python worlds happy with that name.

Reply:
Thanks for the elaboration on the Section and other hoc objects. It's been helpful!